src/mbgl/renderer/layers/render_raster_layer.cpp

```diff
@@ -73,7 +73,16 @@ static std::array<float, 3> spinWeights(float spin) {
     return spin_weights;
 }
 
-void RenderRasterLayer::render(PaintParameters& parameters, RenderSource* source) {
+void RenderRasterLayer::prepare(const LayerPrepareParameters& params) {
+    RenderLayer::prepare(params);
+    auto* imageSource = params.source->as<RenderImageSource>();
+    if (imageSource && imageSource->isLoaded()) {
+        assert(imageSource->isEnabled());
+        imageData = std::make_unique<ImageLayerRenderData>(imageSource->sharedData);
```

Comment:
This can be just a copy of `ImageLayerRenderData`.

Reply:
right, but using unique_ptr enables forward declaration of `ImageLayerRenderData` in header..

Reply:
So, we need `unique_ptr` that holds `shared_ptr`'s just to have forward declaration, correct? Maybe we can just include header with `ImageLayerRenderData` definition.

Reply:
two `shared_ptr` which is 4 raw pointers :smile: Ok, I'll make it optional instead..